packages/common/nbstore/src/storage/doc.ts

```diff
@@ -88,44 +94,62 @@ export interface DocStorage extends Storage {
   ): () => void;
 }
 
-export abstract class DocStorageBase<
-    Opts extends DocStorageOptions = DocStorageOptions,
-  >
-  extends StorageBase<Opts>
+export abstract class DocStorageBase<Opts = {}>
+  extends StorageBase
   implements DocStorage
 {
+  readonly isReadonly: boolean = false;
   private readonly event = new EventEmitter2();
   override readonly storageType = 'doc';
   protected readonly locker: Locker = new SingletonLocker();
+  protected readonly spaceId = this.options.id;
 
-  async getDoc(docId: string) {
-    await using _lock = await this.lockDocForUpdate(docId);
-
-    const snapshot = await this.getDocSnapshot(docId);
-    const updates = await this.getDocUpdates(docId);
-
-    if (updates.length) {
-      const { timestamp, bin, editor } = await this.squash(
-        snapshot ? [snapshot, ...updates] : updates
-      );
-
-      const newSnapshot = {
-        spaceId: this.spaceId,
-        docId,
-        bin,
-        timestamp,
-        editor,
-      };
-
-      await this.setDocSnapshot(newSnapshot, snapshot);
-
-      // always mark updates as merged unless throws
-      await this.markUpdatesMerged(docId, updates);
-
-      return newSnapshot;
+  constructor(protected readonly options: Opts & DocStorageOptions) {
+    super();
+    if (options.readonlyMode) {
+      this.isReadonly = true;
     }
+  }
 
-    return snapshot;
+  async getDoc(docId: string) {
+    const lock = this.isReadonly
+      ? undefined
+      : await this.lockDocForUpdate(docId);
+
+    try {
+      const snapshot = await this.getDocSnapshot(docId);
+      const updates = await this.getDocUpdates(docId);
+
+      if (updates.length) {
+        const { timestamp, bin, editor } = await this.squash(
+          snapshot ? [snapshot, ...updates] : updates
+        );
+
+        const newSnapshot = {
+          spaceId: this.spaceId,
+          docId,
+          bin,
+          timestamp,
+          editor,
+        };
+
+        // if is readonly, we will not set the new snapshot
+        if (!this.isReadonly) {
```

Comment:
readonly should only block update push, the internal merge logic should always work